ldap/servers/slapd/back-ldbm/ldbm_modrdn.c

```diff
@@ -1557,6 +1566,200 @@ ldbm_back_modrdn(Slapi_PBlock *pb)
+        if (parentdns) {
+            for (; parentdns[parentdncomps] != NULL; parentdncomps++)
+                ;
+            slapi_ldap_value_free(parentdns);
+        } else {
+            return -1;
+        }
+    }
+
+    /*
+     * Break down the new superior entry dn into its components.
+     */
+    newsuperiordns = slapi_ldap_explode_dn(slapi_sdn_get_dn(dn_newsuperiordn), 0);
+    if (newsuperiordns) {
+        for (; newsuperiordns[newsuperiordncomps] != NULL; newsuperiordncomps++)
+            ;
+    } else {
+        return -1;
+    }
+
+    /*
+     * Iterate over the child entries renaming them.
+     */
+    child_entry_copies = (struct backentry **)slapi_ch_calloc(sizeof(struct backentry *), nids + 1);
+    for (size_t i = 0; i <= nids; i++) {
+        child_entry_copies[i] = backentry_dup(child_entries[i]);
+    }
+    for (size_t i = 0; retval == 0 && child_entries[i]; i++) {
+        retval = dsentrydn_moddn_rename_child(ptxn, be, li, child_entries[i], &child_entry_copies[i],
+                                              parentdncomps, newsuperiordns,
+                                              newsuperiordncomps);
+    }
+
+    if (0 == retval) {
+        for (size_t i = 0; child_entries[i]; i++) {
+            CACHE_REMOVE(&inst->inst_cache, child_entry_copies[i]);
+            CACHE_RETURN(&inst->inst_cache, &(child_entry_copies[i]));
+        }
+    } else {
+        /* failure */
+        for (size_t i = 0; child_entries[i]; i++) {
+            backentry_free(&(child_entry_copies[i]));
+        }
+    }
+
+    slapi_ch_free((void **)&child_entry_copies);
+    slapi_ldap_value_free(newsuperiordns);
+
+    return retval;
+}
+
+static int32_t
+dsentrydn_modrdn_update(
+    backend *be,
+    const char *olddn,
+    const char *newdn,
+    struct backentry *e,
+    struct backentry **ec,
+    ID id,
+    back_txn *txn)
+{
+    int32_t ret = 0;
+    int32_t cache_rc = 0;
+    ldbm_instance *inst = (ldbm_instance *)be->be_instance_info;
+
+    /* We have the entry, so update id2entry */
+    slapi_entry_attr_set_charptr((*ec)->ep_entry, SLAPI_ATTR_DS_ENTRYDN, newdn);
+    slapi_entry_set_dn((*ec)->ep_entry, (char *)newdn);
+    ret = id2entry_add_ext(be, *ec, txn, 1, &cache_rc);
+    if (cache_rc) {
+        slapi_log_err(SLAPI_LOG_CACHE,
+                      "dsentrydn_modrdn_update",
+                      "Adding %s failed to add to the cache (rc: %d, cache_rc: %d)\n",
+                      slapi_entry_get_dn(e->ep_entry), ret, cache_rc);
+    }
+    if (DBI_RC_RETRY == ret) {
+        /* Retry txn */
+        slapi_log_err(SLAPI_LOG_BACKLDBM, "modrdn_rename_entry_update_indexes",
+                      "id2entry_add deadlock\n");
+        goto out;
+    }
+    if (ret != 0) {
+        slapi_log_err(SLAPI_LOG_ERR, "modrdn_rename_entry_update_indexes",
+                      "id2entry_add failed, err=%d\n",
+                      ret);
+        goto out;
+    }
+    id = e->ep_id;
```

Comment:
If this function should return 'id' then it should be '*id = e->ep_id". But I have the feeling 'id' is a leftover of intermediate version and could be removed.

Reply:
correct